### ISSUE TITLE
Is there something yet to implement?

### ISSUE THREAD

**Issue by SlawekNowy** [nosoop/SM-TFEconData#4]

**Comment by nosoop** [comment]
In terms of matching TF2 Item DB and TF2ItemsInfo, it's missing the ability to read quality colors and unusual effects.  Proper paint kit support would be nice to have as well, though it's handled separately from the text-based schema.

It's usable enough for most use cases.  Provided you're not waiting for those particular things, the rest of the functionality won't be changed in the future (as long as the TF2 team doesn't introduce any breaking changes).

**Comment by SlawekNowy** [comment]
On the case of paintcans:
I just browsed the schema and sure enough there are tags for colors.
General cans have one attribute, teambased have two to be applied to the item.
These are "set_item_tint_rgb" (on all) and "set_item_tint_rgb2"(on teambased only)

**Comment by SlawekNowy** [comment]
As of weapon decaorations, those released before Jungle Inferno are separate items, and those after
might have static atribute "paintkit_proto_def_index" set.

**Comment by SlawekNowy** [comment]
They do have it but, it doesn't seen to be visible

**Comment by nosoop** [comment]
That's all correct.  However, the whole point of the plugin is to read that information from memory.

Paints are trivial to extract with a filtered `TF2Econ_GetItemList` call on paint items, as you've mentioned.  Weapon skins can also be parsed out of existing item definitions to some extent.

Things like `TF2II_GetQualityColor()` (which is what I was referring to for quality colors) aren't exposed in the schema at all; the values are directly in the include file.

I'm not going to add native support for those in that way; plugin authors can write their own convenience methods to do so.  I'm open to providing a `tf_econ_data_stocks` include that provides such abstractions if desired.

Regarding weapon skins, the approach to implement them would be to identify the layout of the `CPaintKit*` game classes and figure out how to extract them from the `CTFItemSchema` instance, if that's possible at all.

I've just pushed an update out to get information on unusual effects, so as far as feature parity with the other two plugins go, this library is complete.

**Comment by SlawekNowy** [comment]
What I found out: https://gist.github.com/SlawekNowy/d46971eee652b0af5540adbe2d7c43e7

**Comment by nosoop** [comment]
Appreciate the efforts.  I actually found all of that a while back myself when I had a lot more interest in covering that stuff, but it doesn't help with reading the contents from server memory.  At that point, you'll need Ghidra, IDA, or another software disassembler and a good understanding of x86 assembly.

The protobuf definition and data files themselves don't tell you that the paintkits are managed by a new `CProtoBufScriptObjectDefinitionManager` class, as opposed to the `CEconItemSchema` that the item schema used, nor does it tell us how paintkits are stored (which is with an `std::vector`-like container which isn't the `CUtlVector` containers Valve normally uses).

Looking into these new things take experience / time to figure out, and the reality is I have a few other projects that better compensate me for my work.

Out of curiosity, is there any particular reason for your interest in this project?  I can't say I recognize your name on AlliedModders, and there's no SourceMod projects to associate you with.

**Comment by SlawekNowy** [comment]
I'm just an "end-user" who is curious how things work.
I was looking for TF2ItemsInfo replacement - tf2idb, and that plugin has tendency of crashing. I raised an issue there, but so far no response.

Then I found this repo.

Why I search for those in the first place. I use minmodels, TF2-GiveItems plugin which I use doesn't have implemented minmodels. TF2ItemsInfo is dead, tf2idb crashes, this has no compat layer.

**Comment by nosoop** [comment]
I see.  The minmodel support is dependent on your weapon granting plugin; it requires that the weapon being granted preserves static attributes (and subsequently overwriting the other attributes at runtime).

This is very much a client-side limitation, as I don't believe you can apply only the `min_viewmodel_offset` attribute at runtime with TF2Attributes; the full vector representation isn't networked.  Blame Valve for making it an attribute instead of a weapon property, really.

This plugin doesn't do any weapon granting on its own.

A compatibility layer is outside the scope of this repository.  I've had a partial one sitting around, but it's never been a priority since I've ported my own plugins, so I don't gain anything from writing one.

**Comment by SlawekNowy** [comment]
About the paintkits I suspect that's all the server does and results are passed to CWeaponSkinProxy on client.

**Comment by nosoop** [comment]
It parses the protobuf data; I'd imagine it's sitting around in memory somewhere.  The game just doesn't access them.

On further inspection one can call `IProtoBufScriptObjectDefinition::GetName()` on the `CPaintKitDefinition` to get the internal name.  The virtual `CPaintKitDefinition::GetDisplayName()` is overwritten to display the `loc_desctoken` property instead, so a virtual call is out of the question here &mdash; it has to be a signature.  Windows support would be problematic since the functions aren't all that long.

I can't really justify the addition / maintenance just to get internal names on paintkits; some other plugin can do it if they want (which is why `TF2Econ_GetPaintKitDefinitionAddress` is available for plugins to use).

As an aside, [I did release the part of the work I've done on a compatibility shim here][econ-compat], in case you're still looking for one.

[econ-compat]: https://github.com/nosoop/SM-TFEconDataCompat
